### PR TITLE
dcm2niix: add support for configuring optional compile-time modules 

### PR DESCRIPTION
**Pull request by rbreslow**
###### Description of changes

Right now, the dcm2niix package produces a single binary without support for dcm2niibatch, JPEG-LS, OpenJPEG, or an alternative zlib library. The [official build process](https://github.com/rordenlab/dcm2niix/blob/002ebcdb9b2a87de7b883e9ddada3963a1cc2327/.travis.yml#L29) for dcm2niix configures all these options (minus batch support). I was motivated to make these changes so my organization could process JPEG 2000 imagery using the Nix binary.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by rbreslow** [comment]
Result of `nixpkgs-review pr 165477` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>dcm2niix</li>
  </ul>
</details>

File: pkgs/applications/science/biology/dcm2niix/default.nix [hunk 1]
```diff
@@ -17,7 +24,23 @@ stdenv.mkDerivation rec {
   };
 
   nativeBuildInputs = [ cmake git ];
-  buildInputs = [ libyamlcpp ];
+  buildInputs =
+    lib.optional batchVersion [ libyamlcpp ]
```
Comment (rbreslow) [inline pkgs/applications/science/biology/dcm2niix/default.nix:28]: This library is only used if you enable batch processing support:

https://github.com/rordenlab/dcm2niix/blob/003f0d19f1e57b0129c9dcf3e653f51ca3559028/SuperBuild/SuperBuild.cmake#L88-L107.

File: pkgs/applications/science/biology/dcm2niix/default.nix [hunk 2]
```diff
@@ -17,7 +24,23 @@ stdenv.mkDerivation rec {
...
+      "-DBATCH_VERSION=ON"
+      "-DYAML-CPP_DIR=${libyamlcpp}/lib/cmake/yaml-cpp"
+    ] ++ lib.optional withJpegLs [
+      "-DUSE_JPEGLS=ON"
+    ] ++ lib.optional withOpenJpeg [
+      "-DUSE_OPENJPEG=ON"
+      "-DOpenJPEG_DIR=${openjpeg}/lib/${openjpeg.pname}-${lib.versions.majorMinor openjpeg.version}"
+    ] ++ lib.optional withSystemZlib [
+      "-DZLIB_IMPLEMENTATION=System"
+    ];
```
Comment (rbreslow) [inline pkgs/applications/science/biology/dcm2niix/default.nix:43]: The "official" build process uses a fork of Cloudflare's zlib fork:

- https://github.com/rordenlab/dcm2niix/blob/002ebcdb9b2a87de7b883e9ddada3963a1cc2327/.travis.yml#L29
- https://github.com/rordenlab/dcm2niix/blob/2bf2e482aec8e9959c6bd8e833cdccba3607c617/SuperBuild/External-CLOUDFLARE-ZLIB.cmake

Rather than fall back to the included miniZ zlib implementation, I thought it'd be best to bridge the gap and use the zlib included with NixOS.

File: pkgs/applications/science/biology/dcm2niix/default.nix [hunk 3]
```diff
@@ -17,7 +24,23 @@ stdenv.mkDerivation rec {
   };
 
   nativeBuildInputs = [ cmake git ];
-  buildInputs = [ libyamlcpp ];
+  buildInputs =
+    lib.optional batchVersion [ libyamlcpp ]
+    ++ lib.optional withOpenJpeg [ openjpeg openjpeg.dev ]
+    ++ lib.optional withSystemZlib [ zlib ];
```
Comment (SuperSandro2000) [inline pkgs/applications/science/biology/dcm2niix/default.nix:30]:
```suggestion
  buildInputs = [ zlib ]
    ++ lib.optional batchVersion libyamlcpp
    ++ lib.optionals withOpenJpeg [ openjpeg openjpeg.dev ];
```

File: pkgs/applications/science/biology/dcm2niix/default.nix [hunk 4]
```diff
@@ -1,8 +1,15 @@
...
 , cmake
-, libyamlcpp
 , git
+, openjpeg
+, libyamlcpp
+, zlib
+, batchVersion ? false
+, withJpegLs ? true
+, withOpenJpeg ? true
+, withSystemZlib ? true
```
Comment (SuperSandro2000) [inline pkgs/applications/science/biology/dcm2niix/default.nix:12]:
```suggestion
```
We always wan to do that.

File: pkgs/applications/science/biology/dcm2niix/default.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/applications/science/biology/dcm2niix/default.nix:43]: Oh, if there is an actual use case for the vendored zlib we should name the option something like ``withCloudflareZlib`` or something that indicates the better performance.

Comment (SuperSandro2000) [inline pkgs/applications/science/biology/dcm2niix/default.nix:43]:
```suggestion
  cmakeFlags = [
    "-DZLIB_IMPLEMENTATION=System"
  ] ++ lib.optionals batchVersion [
    "-DBATCH_VERSION=ON"
    "-DYAML-CPP_DIR=${libyamlcpp}/lib/cmake/yaml-cpp"
  ] ++ lib.optionals withJpegLs [
    "-DUSE_JPEGLS=ON"
  ] ++ lib.optionals withOpenJpeg [
    "-DUSE_OPENJPEG=ON"
    "-DOpenJPEG_DIR=${openjpeg}/lib/${openjpeg.pname}-${lib.versions.majorMinor openjpeg.version}"
  ];
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/applications/science/biology/dcm2niix/default.nix (hunk 2 above)
Comment (rbreslow) [reply pkgs/applications/science/biology/dcm2niix/default.nix]: By default., `dcm2niix` is compiled with the vendored zlib. I don't want to change the default behavior. As an optional flag, CMake will fetch a fork of Cloudflare's zlib. However, there's no Nix package for that. System zlib is a decent middle ground, but I don't want to make it a default since it hasn't been tested upstream. What are your thoughts?

File: pkgs/applications/science/biology/dcm2niix/default.nix [hunk 5]
```diff
@@ -17,7 +24,22 @@ stdenv.mkDerivation rec {
   };
 
   nativeBuildInputs = [ cmake git ];
-  buildInputs = [ libyamlcpp ];
+  buildInputs = with lib;
```
Comment (SuperSandro2000) [inline pkgs/applications/science/biology/dcm2niix/default.nix:27]: I don't think we need with here. lib is so short and nice that we can just write it out.

File: pkgs/applications/science/biology/dcm2niix/default.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/applications/science/biology/dcm2niix/default.nix]: Distros almost always devendor libaries because they are more often than not horrible out of date and though miss security patches which you also don't want to apply in several places.

> However, there's no Nix package for that.

We can add one for it. As far as I understood it, it should be a drop in replacement hence we also don't need to thorough testing.

Comment (rbreslow) [reply pkgs/applications/science/biology/dcm2niix/default.nix]: > Distros almost always devendor libraries because they are more often than not horrible out of date and though miss security patches which you also don't want to apply in several places.

This makes sense—100%. I suppose I fear that Miniz is a separate codebase than zlib. It's an alternative library that implements the zlib RFC. So, we're assuming liability by devendoring the library and switching to NixOS's zlib implementation. There could be subtle nuances between the Miniz zlib, NixOS's zlib, and this Cloudflare fork supported by the upstream repo. And, dcm2niix doesn't have a test suite, so there's no low-effort way to verify that things still work.

BTW, I appreciate the discussion and your time 🙂.

It sounds like our options are:

1) Assume the liability. Switch to NixOS's zlib. Hope that it's compatible with dcm2niix and users of the package don't encounter any nuance.
2) Bring in Cloudflare's Zlib to Nixpkgs and add an option called `withCloudflareZlib`. However, dcm2niix uses a frozen fork of Cloudflare's Zlib, so incompatibilities could crop up if we packaged a newer version in the distro.

I'm leaning towards #1 to get this out the door, but I'm also nervous about using this tool in our production environment with that enabled by default.

Comment (rbreslow) [reply pkgs/applications/science/biology/dcm2niix/default.nix]: I went with option one.

**Comment by neurolabusc** [comment]
1. It is worth noting that if you use cmake you can use `cmake -LA` to list options, which reveals how to turn on and off specific features, such as building the batch version. The suggested options are `-DZLIB_IMPLEMENTATION=Cloudflare -DUSE_JPEGLS=ON -DUSE_OPENJPEG=ON` The zlib library alternatives are miniz, system or Cloudflare. For JPEG 2000, I would use OpenJPEG instead of Jasper as the latter has known issues with some 16-bit images. 

```
BATCH_VERSION:BOOL=OFF
USE_JASPER:BOOL=OFF
USE_JPEGLS:BOOL=OFF
USE_OPENJPEG:BOOL=OFF
USE_STATIC_RUNTIME:BOOL=ON
USE_TURBOJPEG:BOOL=OFF
```

2. The make command provides far fewer options that cmake - it is really for development not deployment. You can build aarch64-darwin and  x86_64-darwin on a single Mac and use Lipo to create a universal binary. If you ask, I can send you the script that allows you to notarize the resulting binary (you will need an Apple Developer ID and generate an Apple App Specific password):

```
g++ -O3 -sectcreate __TEXT __info_plist Info.plist -I.  main_console.cpp nii_foreign.cpp nii_dicom.cpp jpg_0XC3.cpp ujpeg.cpp nifti1_io_core.cpp nii_ortho.cpp nii_dicom_batch.cpp  -o dcm2niixX86 -DmyDisableOpenJPEG -target x86_64-apple-macos10.12 -mmacosx-version-min=10.12
g++ -O3 -sectcreate __TEXT __info_plist Info.plist -I.  main_console.cpp nii_foreign.cpp nii_dicom.cpp jpg_0XC3.cpp ujpeg.cpp nifti1_io_core.cpp nii_ortho.cpp nii_dicom_batch.cpp  -o dcm2niixARM -DmyDisableOpenJPEG -target arm64-apple-macos11 -mmacosx-version-min=11.0
# Create the universal binary.
strip ./dcm2niixARM; strip ./dcm2niixX86
lipo -create -output ${APP_NAME} dcm2niixARM dcm2niixX86
```

**Comment by rbreslow** [comment]
@neurolabusc, thank you for popping into the thread. I added support for your Cloudflare zlib fork in 438489a. 

**Some considerations:**

- The Cloudflare zlib fork only has a `gcc.amd64` branch, so I'm not sure what will happen if someone tries to compile this package with `withCloudflareZlib` on another architecture.
- I'm not using the yaml-cpp and OpenJPEG sources you defined since there are variants packaged for NixOS. However, it looks like NixOS [provides](https://github.com/NixOS/nixpkgs/blob/634141959076a8ab69ca2cca0f266852256d79ee/pkgs/development/libraries/openjpeg/default.nix) OpenJPEG 2.4.0, as opposed to 2.1.

Let me know what you think 🙂.

File: pkgs/applications/science/biology/dcm2niix/default.nix [hunk 6]
```diff
@@ -1,10 +1,24 @@
...
 }:
 
+let
+  cloudflareZlib = fetchFromGitHub {
+    owner = "ningfei";
+    repo = "zlib";
+    rev = "gcc.amd64";
+    sha256 = "sha256-qySFwY0VI2BQLO2XoCZeYshXEDnHh6SmJ3MvcBUROWU=";
+  };
+in
```
Comment (rbreslow) [inline pkgs/applications/science/biology/dcm2niix/default.nix:21]: @neurolabusc, we have to fetch these sources outside of CMake because the build environment is isolated from the network. NixOS would consider a network request during the build process "impure."

**Comment by neurolabusc** [comment]
I will let @ningfei comment, his [Cloudflare zlib](https://github.com/ningfei/zlib) fork may not have all of the latest ARM updates as the main branch. However, it does include some ARM SIMD tuning. You can see this when you build the tool as well with his cmake script:

```
git clone --branch development https://github.com/rordenlab/dcm2niix.git
cd dcm2niix
mkdir build && cd build
cmake -DZLIB_IMPLEMENTATION=Cloudflare -DUSE_JPEGLS=ON ..
make

...
-- Performing Test ARM_CRC
-- Performing Test ARM_CRC - Success
-- Found ZLIB: /Users/chrisrorden/neuro/dcm2niix/build/lib/libz.a (found version "1.2.8") 
...

file ./bin/dcm2niix
./bin/dcm2niix: Mach-O 64-bit executable arm64

./bin/dcm2niix 
Chris Rorden's dcm2niiX version v1.0.20220322  (JP-LS:CharLS) Clang12.0.5 ARM (64-bit MacOS)
...
  -z : gz compress images (y/o/i/n/3, default n) [y=pigz, o=optimal pigz, i=internal:zlib, n=no, 3=no,3D]
```

By the way, the proprietary system zlib that comes with the Apple M1 has been specially tuned by Apple, and it clearly uses a lot of the same tricks as CloudFlare. While CloudFlare dramatically outperforms the system zlib on x86-64 MacOS, Windows and Linux distributions, on ARM both the Apple system zlib and CloudFlare are blazing fast.

@ningfei may have a better memory of the OpenJPEG saga, but as I recall if you install the (outstanding) gdcm library from source it will install its own custom OpenJPEG library that is not compatible with the standard library but indistinguishable to compilers of other code. Therefore, I think we have to be very careful about using pre-existing copies of that library.

**Comment by rbreslow** [comment]
@SuperSandro2000 can you take another pass at this PR? I think it'd be valuable to incorporate any feedback from @ningfei, but I'd also like to get something merged so that we can have any build of dcm2niix with support for these additional modules.

**Comment by ningfei** [comment]
Hi, sorry for the late reply! For Cloudflare zlib, the branch name might be misleading. It does support other architectures, as long as it is 64bit if I remember correctly. Not all ARM related optimizations were included, but we can for sure consider add more. As for yaml-cpp and OpenJPEG, you can use the libraries shipped by NixOS. There used to be compatibility issue for OpenJPEG. But as I just tested, at least the compilation worked for system libopenjp2 v2.3. The only small downside is that static linking won't work if you link against system libopenjp2.

**Comment by ningfei** [comment]
...add better do some tests with the generated dcm2niix to see if it OpenJPEG v2.4 works properly. I think dcm_qa has related test sample in it? @neurolabusc

**Comment by rbreslow** [comment]
> ...add better do some tests with the generated dcm2niix to see if it OpenJPEG v2.4 works properly. I think dcm_qa has related test sample in it?

If the script in [dcm_qa](https://github.com/neurolabusc/dcm_qa) is a solid, reproducible test case, I can add an automated test that we can trigger any time Hydra (NixOS CI system) rebuilds the package. Also, I see there are a lot of "QA" repositories on Chris's GitHub. If you can provide a list of the current ones, we can run through all of them. Although, I think it'd be worth doing that in a separate pull request (which I'm happy to open).

**Comment by rbreslow** [comment]
> The only small downside is that static linking won't work if you link against system libopenjp2.

Nix discourages static linking. At a high level, Nix caches different shared library versions in the [Nix store](https://nixos.org/manual/nix/stable/introduction.html) and reuses them across all your packages. Folks shouldn't be consuming this build of dcm2niix outside of Nix (although it's possible for Nix to produce a statically-linked binary for GitHub releases, etc.).

**Comment by neurolabusc** [comment]
@rbreslow the basic [dcm_qa includes a JPEG2000 test](https://github.com/neurolabusc/dcm_qa/tree/master/In/Orientation/axmb/AxInt36mb). Each commit of dcm2niix is tested against dcm_qa (Siemens), dcm_qa_nih (Siemens, GE) and dcm_qa_uih (UIH). 

You can use Github's search for a full list of [dcm_qa_*](https://github.com/neurolabusc?tab=repositories&q=dcm_qa&type=&language=&sort=). Most of these were to demonstrate edge cases and I only test these on a very irregular basis.

File: pkgs/applications/science/biology/dcm2niix/default.nix [hunk 7]
```diff
@@ -1,10 +1,24 @@
...
+, withJpegLs ? true
+, withOpenJpeg ? true
+, withCloudflareZlib ? true
 }:
 
+let
+  cloudflareZlib = fetchFromGitHub {
+    owner = "ningfei";
+    repo = "zlib";
+    rev = "gcc.amd64";
```
Comment (SuperSandro2000) [inline pkgs/applications/science/biology/dcm2niix/default.nix:18]: We can't track a branch. As soon as a commit is pushed there the hash is wrong. Please track a commit or tag.

Comment (rbreslow) [reply pkgs/applications/science/biology/dcm2niix/default.nix:18]: > We can't track a branch.

Is this a Nixpkgs contribution policy? 

> As soon as a commit is pushed there the hash is wrong.

Yes, I know this. However, the upstream project also pulls the latest commit from the `gcc.amd64` branch. So if there is an update to that branch, I want the Nix package build to fail as a prompt to update the hash. That way, we'll stay in sync with the upstream version of the dependency.

Before I proceed with any changes, please share your thoughts on this @SuperSandro2000.

Comment (SuperSandro2000) [reply pkgs/applications/science/biology/dcm2niix/default.nix:18]: > Is this a Nixpkgs contribution policy?

I don't know but as soon as someone pushes to that branch the has is wrong which is not a reproducible build in any way.

> I want the Nix package build to fail as a prompt to update the hash.

Thats not how packages are supposed to work. We use update scripts and bots for this.

Comment (rbreslow) [reply pkgs/applications/science/biology/dcm2niix/default.nix]: Please take another look @SuperSandro2000.

Comment (rbreslow) [reply pkgs/applications/science/biology/dcm2niix/default.nix]: Hey @SuperSandro2000, this has been hanging out for a long time, can you take a look or remove yourself as a reviewer?

**Comment by rbreslow** [comment]
Thank you, all!
